api/v1alpha2/jenkins_webhook.go

```diff
@@ -41,21 +49,140 @@ var _ webhook.Validator = &Jenkins{}
+}
+type Version struct {
+	FirstVersion string `json:"firstVersion"`
+	LastVersion  string `json:"lastVersion"`
+}
+
+const APIURL string = "https://plugins.jenkins.io/api/plugin/"
+
+func MakeSemanticVersion(version string) string {
+	version = "v" + version
+	return semver.Canonical(version)
+}
+
+func CompareVersions(firstVersion string, lastVersion string, pluginVersion string) bool {
+	firstSemVer := MakeSemanticVersion(firstVersion)
+	lastSemVer := MakeSemanticVersion(lastVersion)
+	pluginSemVer := MakeSemanticVersion(pluginVersion)
+	if semver.Compare(pluginSemVer, firstSemVer) == -1 || semver.Compare(pluginSemVer, lastSemVer) == 1 {
+		return false
+	}
+	return true
+}
+
+func CheckSecurityWarnings(pluginName string, pluginVersion string) (bool, error) {
+	jenkinslog.Info("checking security warnings", "plugin: ", pluginName)
+	pluginURL := APIURL + pluginName
+	client := &http.Client{
+		Timeout: time.Second * 30,
+	}
+	request, err := http.NewRequest("GET", pluginURL, nil)
+	if err != nil {
+		return false, err
+	}
+	request.Header.Add("Accept", "application/json")
+	request.Header.Add("Content-Type", "application/json")
+	response, err := client.Do(request)
+	if err != nil {
+		return false, err
+	}
+	defer response.Body.Close()
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return false, err
+	}
+	securityWarnings := Warnings{}
+
+	jsonErr := json.Unmarshal(bodyBytes, &securityWarnings)
+	if jsonErr != nil {
+		return false, err
+	}
+
+	jenkinslog.Info("Validate()", "warnings", securityWarnings)
+
+	for _, warning := range securityWarnings.Warnings {
+		for _, version := range warning.Versions {
+			firstVersion := version.FirstVersion
+			lastVersion := version.LastVersion
+			if len(firstVersion) == 0 {
+				firstVersion = "0" // setting default value in case of empty string
+			}
+			if len(lastVersion) == 0 {
+				lastVersion = pluginVersion // setting default value in case of empty string
+			}
+
+			if CompareVersions(firstVersion, lastVersion, pluginVersion) {
+				jenkinslog.Info("security Vulnerabilities detected", "message", warning.Message, "Check security Advisory", warning.URL)
+				return true, nil
+			}
+		}
+	}
+
+	return false, nil
+}
+
+func Validate(r Jenkins) error {
+	basePlugins := plugins.BasePlugins()
+
+	for _, plugin := range basePlugins {
+		name := plugin.Name
+		version := plugin.Version
+		hasWarnings, err := CheckSecurityWarnings(name, version)
+		if err != nil {
+			return err
+		}
+		if hasWarnings {
+			errormsg := "Security Vulnerabilities detected in base plugin:" + name
+			return errors.New(errormsg)
+		}
+	}
+
+	for _, plugin := range r.Spec.Master.Plugins {
+		name := plugin.Name
+		version := plugin.Version
+		hasWarnings, err := CheckSecurityWarnings(name, version)
+		if err != nil {
+			return err
+		}
+		if hasWarnings {
+			errormsg := "Security Vulnerabilities detected in the user defined plugin: " + name
+			return errors.New(errormsg)
+		}
+	}
+
```

Comment:
If you look closely, you are returning with the first warning you encountered, so if there's more than 1 warning you're skipping it. It would be better if you were appending warnings and leave only when you iterate through all the plugins. Right now you wouldn't even start checking user plugins. Please add some unit tests. For those cases too, eg. 2 warnings in user plugins and 2 warnings in user plugins and check if you've gotten them all.